### PR TITLE
[tune] Add `save` and `restore` methods for searchers that were missing it & test

### LINKED ISSUE #18399: [tune] BOHB, Ax, Dragonfly searchers are missing save and restore functionality
### What is the problem?

BOHB, Ax, Dragonfly searchers are missing save and restore functionality.

Followup from https://github.com/ray-project/ray/issues/18271

### Reproduction (REQUIRED)
Please provide a short code snippet (less than 50 lines if possible) that can be copy-pasted to reproduce the issue. The snippet should have **no external library dependencies** (i.e., use fake or mock data / environments):

If the code snippet cannot be run by itself, the issue will be closed with "needs-repro-script".

- [ ] I have verified my script runs in a clean environment and reproduces the issue.
- [ ] I have verified the issue also occurs with the [latest wheels](https://docs.ray.io/en/master/installation.html).

### PR DESCRIPTION
**Pull request by Yard1**
## Why are these changes needed?

Adds `save` and `restore` methods for Ax, BOHB and Dragonfly searchers, as well as a simple unit test to check if it's working for every searcher. Enables the existing, more in-depth test for save and restore for all searchers.

Also updates existing save and restore methods to be more robust, and add a `random_state_seed` argument to `DragonflySearch`, working similarly to the one in `HEBOSearch`.

## Related issue number

Closes https://github.com/ray-project/ray/issues/18399

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/tune/tests/test_searchers.py [hunk 1]
```diff
@@ -331,6 +334,170 @@ def testHEBO(self):
...
+        searcher.suggest("1")
+        searcher.suggest("2")
+        searcher.on_trial_complete("1", {
+            self.metric_name: 1,
+            "config/a": 1.0,
+            "time_total_s": 1
+        })
+
+        searcher.save(self.checkpoint_path)
+        searcher.restore(self.checkpoint_path)
```
Comment (krfricke) [inline python/ray/tune/tests/test_searchers.py:372]: Use new searcher object

**Comment by krfricke** [comment]
Let's add the warmstart tests here, too

**Review by krfricke — APPROVED** [review]
Looks good - quick question, what about Dragonfly warm start test?

**Comment by Yard1** [comment]
@krfricke It has been turned on, and warm start & restore tests have been added for all searchers that were missing them (so now all searchers are being tested).

**Pull request merged**